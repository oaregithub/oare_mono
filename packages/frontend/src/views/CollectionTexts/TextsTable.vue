<template>
  <v-data-table
    :headers="headers"
    :items="loading ? [] : texts"
    :loading="loading"
    :options.sync="searchOptions"
    :server-items-length="totalTexts"
    :footer-props="{
      'items-per-page-options': [10, 25, 50, 100],
    }"
  >
    <template v-slot:[`item.name`]="{ item }">
      <router-link v-if="item.hasEpigraphy" :to="`/epigraphies/${item.uuid}`">{{
        item.name
      }}</router-link>
      <span v-else>{{ item.name }}</span>
    </template>
  </v-data-table>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  computed,
  watch,
  PropType,
} from '@vue/composition-api';
import { CollectionResponse, CollectionText } from '@/types/collections';
export default defineComponent({
  props: {
    loading: {
      type: Boolean,
      default: false,
    },
    texts: {
      type: Array as PropType<CollectionText[]>,
      required: true,
    },
    totalTexts: {
      type: Number,
      required: true,
    },
    page: {
      type: Number,
      default: 1,
    },
    rows: {
      type: Number,
      default: 10,
    },
  },
  setup(props, { emit }) {
    const headers = ref([
      {
        text: 'Text Name',
        value: 'name',
      },
    ]);

    const searchOptions = ref({
      page: props.page,
      itemsPerPage: props.rows,
    });

    watch(
      () => props.page,
      () => (searchOptions.value.page = props.page)
    );

    watch(
      () => searchOptions.value.page,
      () => {
        emit('update:page', searchOptions.value.page);
      }
    );

    watch(
      () => searchOptions.value.itemsPerPage,
      () => {
        emit('update:rows', searchOptions.value.itemsPerPage);
      }
    );

    return {
      headers,
      searchOptions,
    };
  },
});
</script>

<style></style>
