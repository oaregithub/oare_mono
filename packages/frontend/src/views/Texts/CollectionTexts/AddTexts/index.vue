<template>
  <OareContentView
    :title="`Add New Text to ${collectionName}`"
    :loading="loading"
  >
    <v-stepper v-model="step">
      <v-stepper-header>
        <v-stepper-step :complete="step > 1" step="1">
          Text Info
        </v-stepper-step>
        <v-divider />
        <v-stepper-step :complete="step > 2" step="2">
          Upload Photos
        </v-stepper-step>
        <v-divider />
        <v-stepper-step :complete="step > 3" step="3"> Editor </v-stepper-step>
        <v-divider />
        <v-stepper-step :complete="step > 4" step="4">
          Lexical Info
        </v-stepper-step>
        <v-divider />
        <v-stepper-step step="5"> Final Preview </v-stepper-step>
      </v-stepper-header>

      <v-stepper-items>
        <v-stepper-content step="1">
          <text-info-set
            @update-text-info="setTextInfo"
            @step-complete="stepOneComplete = $event"
          />
          <stepper-button
            :showBackButton="false"
            :blockContinue="!stepOneComplete"
            blockContinueText="To continue, please include a text name. Please also include a prefix and number for at least one of the following: Excavation, Museum, Publication"
            @next="next"
          />
        </v-stepper-content>

        <v-stepper-content step="2">
          <add-photos
            v-if="step >= 2"
            @update-photos="setPhotos"
            @step-complete="stepTwoComplete = $event"
          />
          <stepper-button
            :blockContinue="!stepTwoComplete"
            blockContinueText="To continue, each photo needs to have a side and view selected. Please make sure you also have uploaded photos for each selector."
            @next="next"
            @previous="previous"
          />
        </v-stepper-content>

        <v-stepper-content step="3">
          <add-text-editor
            @update-editor-content="setEditorContent"
            @step-complete="stepThreeComplete = $event"
          />
          <stepper-button
            :blockContinue="!stepThreeComplete"
            blockContinueText="To continue, there must be at least one side. No empty sides,
            columns, or rows are allowed. All signs and markup must be valid."
            :continueAction="buildTables"
            @next="next"
            @previous="previous"
          />
        </v-stepper-content>

        <v-stepper-content step="4">
          <connect-discourse
            v-if="step >= 4"
            :epigraphicUnits="epigraphyDetails.units"
            :discourseRows="createTextTables ? createTextTables.discourses : []"
            :manualDiscourseSelections="manuallySelectedDiscourses"
            @update-discourse-rows="updateDiscourseRows($event)"
            @update-manual-selections="updateManualSelections($event)"
          />
          <stepper-button @next="next" @previous="previous" />
        </v-stepper-content>

        <v-stepper-content step="5">
          <final-preview
            v-if="step >= 5"
            :epigraphyUnits="epigraphyDetails"
            :photoUrls="photoUrls"
            :localDiscourseInfo="
              createTextTables ? createTextTables.discourses : []
            "
          />
          <stepper-button
            continueButtonText="Submit"
            :continueAction="createText"
            @previous="previous"
            @next="pushToText"
          />
        </v-stepper-content>
      </v-stepper-items>
    </v-stepper>
  </OareContentView>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  onMounted,
  onBeforeMount,
  onBeforeUnmount,
  computed,
  ComputedRef,
} from '@vue/composition-api';
import sl from '@/serviceLocator';
import AddTextEditor from './Editor/AddTextEditor.vue';
import TextInfoSet from './TextInfo/TextInfoSet.vue';
import AddPhotos from './Photos/AddPhotos.vue';
import ConnectDiscourse from './Discourse/ConnectDiscourse.vue';
import EpigraphyView from '@/views/Texts/EpigraphyView/index.vue';
import FinalPreview from '@/views/Texts/CollectionTexts/AddTexts/Preview/FinalPreview.vue';
import StepperButton from './components/StepperButton.vue';
import {
  AddTextEditorContent,
  AddTextInfo,
  TextPhoto,
  EpigraphyResponse,
  CreateTextTables,
  TextDiscourseRow,
  TextPhotoWithName,
} from '@oare/types';
import { convertTablesToUnits } from './utils/convertTablesToUnits';
import { createNewTextTables } from './utils/buildTables';
import { addNamesToTextPhotos } from './utils/photos';

export default defineComponent({
  props: {
    collectionUuid: {
      type: String,
      required: true,
    },
  },
  beforeRouteLeave(_to, _from, next) {
    if (!this.isDirty) {
      next();
    } else {
      this.actions.showUnsavedChangesWarning(next);
    }
  },
  components: {
    AddTextEditor,
    TextInfoSet,
    AddPhotos,
    EpigraphyView,
    ConnectDiscourse,
    FinalPreview,
    StepperButton,
  },
  setup(props) {
    const server = sl.get('serverProxy');
    const actions = sl.get('globalActions');
    const router = sl.get('router');

    const collectionName = ref('');
    const step = ref(1);
    const loading = ref(false);

    const stepOneComplete = ref(false);
    const stepTwoComplete = ref(false);
    const stepThreeComplete = ref(false);

    const isDirty = ref(true);

    const textInfo = ref<AddTextInfo>();
    const setTextInfo = (updatedTextInfo: AddTextInfo) => {
      textInfo.value = updatedTextInfo;
    };

    const photos = ref<TextPhoto[]>([]);
    const photosWithName = ref<TextPhotoWithName[]>([]);
    const setPhotos = (updatedPhotos: TextPhoto[]) => {
      photos.value = updatedPhotos;
    };
    const photoUrls = computed(() =>
      photosWithName.value.filter(photo => photo.url).map(photo => photo.url)
    );

    const editorContent = ref<AddTextEditorContent>();
    const setEditorContent = (updatedEditorContent: AddTextEditorContent) => {
      editorContent.value = updatedEditorContent;
    };

    const epigraphyDetails: ComputedRef<EpigraphyResponse> = computed(() => {
      return {
        canWrite: false,
        text: {
          uuid: createTextTables.value ? createTextTables.value.text.uuid : '',
          type: 'logosyllabic',
          name:
            textInfo.value && textInfo.value.textName
              ? textInfo.value.textName
              : '',
          excavationPrefix:
            textInfo.value && textInfo.value.excavationPrefix
              ? textInfo.value.excavationPrefix
              : '',
          excavationNumber:
            textInfo.value && textInfo.value.excavationNumber
              ? textInfo.value.excavationNumber
              : '',
          museumPrefix:
            textInfo.value && textInfo.value.museumPrefix
              ? textInfo.value.museumPrefix
              : '',
          museumNumber:
            textInfo.value && textInfo.value.museumNumber
              ? textInfo.value.museumNumber
              : '',
          publicationPrefix:
            textInfo.value && textInfo.value.publicationPrefix
              ? textInfo.value.publicationPrefix
              : '',
          publicationNumber:
            textInfo.value && textInfo.value.publicationNumber
              ? textInfo.value.publicationNumber
              : '',
        },
        collection: {
          uuid: props.collectionUuid,
          name: collectionName.value,
        },
        cdliNum:
          textInfo.value && textInfo.value.cdliNum
            ? textInfo.value.cdliNum
            : '',
        units: createTextTables.value
          ? convertTablesToUnits(createTextTables.value)
          : [],
        color: '',
        colorMeaning: '',
        discourseUnits: [],
        hasEpigraphy: true,
      };
    });

    onMounted(async () => {
      loading.value = true;
      try {
        collectionName.value = (
          await server.getCollectionInfo(props.collectionUuid)
        ).name;
      } catch {
        actions.showErrorSnackbar(
          'Error loading collection name. Please try again.'
        );
      } finally {
        loading.value = false;
      }
    });

    const next = () => (step.value += 1);
    const previous = () => (step.value -= 1);

    const pushToText = () => {
      if (createTextTables.value) {
        router.push(`/epigraphies/${createTextTables.value.text.uuid}`);
      }
    };

    const createText = async () => {
      if (createTextTables.value) {
        try {
          isDirty.value = false;
          await server.createText(createTextTables.value);
          await Promise.all(
            photosWithName.value.map(photo => server.uploadImage(photo))
          );
        } catch (err) {
          actions.showErrorSnackbar(
            'Error creating text. Please try again',
            err as Error
          );
        }
      }
    };

    const createTextTables = ref<CreateTextTables>();
    const persistentDiscourseStorage = ref<{ [uuid: string]: string | null }>(
      {}
    );
    const manuallySelectedDiscourses = ref<string[]>([]);
    const updateManualSelections = (discourseUuid: string) => {
      manuallySelectedDiscourses.value =
        manuallySelectedDiscourses.value.filter(uuid => uuid !== discourseUuid);
      manuallySelectedDiscourses.value.push(discourseUuid);
    };
    const buildTables = async () => {
      if (textInfo.value && editorContent.value) {
        photosWithName.value = await addNamesToTextPhotos(
          textInfo.value.excavationPrefix,
          textInfo.value.excavationNumber,
          textInfo.value.museumPrefix,
          textInfo.value.museumNumber,
          textInfo.value.publicationPrefix,
          textInfo.value.publicationNumber,
          photos.value
        );

        createTextTables.value = await createNewTextTables(
          textInfo.value,
          editorContent.value,
          persistentDiscourseStorage.value,
          photosWithName.value,
          props.collectionUuid
        );
      }
      if (createTextTables.value) {
        createTextTables.value.discourses.forEach(discourse => {
          persistentDiscourseStorage.value[discourse.uuid] =
            discourse.spellingUuid;
        });
      }
    };

    const updateDiscourseRows = (discourses: TextDiscourseRow[]) => {
      if (createTextTables.value) {
        createTextTables.value = {
          ...createTextTables.value,
          discourses,
        };
        createTextTables.value.discourses.forEach(row => {
          persistentDiscourseStorage.value[row.uuid] = row.spellingUuid;
        });
      }
    };

    const beforeUnloadHandler = (event: BeforeUnloadEvent) => {
      if (isDirty) {
        event.preventDefault();
        event.returnValue = '';
      }
    };

    onBeforeMount(() => {
      window.addEventListener('beforeunload', beforeUnloadHandler);
    });

    onBeforeUnmount(() => {
      window.removeEventListener('beforeunload', beforeUnloadHandler);
    });

    return {
      actions,
      collectionName,
      step,
      loading,
      setTextInfo,
      setEditorContent,
      setPhotos,
      photoUrls,
      epigraphyDetails,
      next,
      previous,
      createTextTables,
      updateDiscourseRows,
      stepOneComplete,
      stepTwoComplete,
      stepThreeComplete,
      buildTables,
      pushToText,
      createText,
      manuallySelectedDiscourses,
      updateManualSelections,
      isDirty,
    };
  },
});
</script>
