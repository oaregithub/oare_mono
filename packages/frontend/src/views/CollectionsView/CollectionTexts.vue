<template>
  <v-progress-linear v-if="loading" indeterminate />
  <OareContentView v-else :title="collectionName">
    <template #header>
      <OareBreadcrumbs :items="breadcrumbItems" />
    </template>
    <v-spacer />
    <v-container>
      <v-row>
        <v-col cols="4" offset="8">
          <v-text-field
            v-model="search"
            label="Search"
            single-line
            hide-details
            clearable
          />
        </v-col>
      </v-row>
      <TextsTable
        :page.sync="page"
        :rows.sync="rows"
        :totalTexts="totalTexts"
        :texts="texts"
        :loading="textsLoading"
      />
    </v-container>
  </OareContentView>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  watch,
  Ref,
  computed,
  onMounted,
} from '@vue/composition-api';
import { CollectionResponse, CollectionText } from '@/types/collections';
import { BreadcrumbItem } from '@/components/base/OareBreadcrumbs.vue';
import TextsTable from './TextsTable.vue';
import { getLetterGroup } from './utils';
import _ from 'underscore';
import router from '@/router';
import serverProxy from '@/serverProxy';

export default defineComponent({
  name: 'CollectionTexts',
  components: {
    TextsTable,
  },
  props: {
    collectionUuid: {
      type: String,
      required: true,
    },
  },

  setup(props, context) {
    const collectionName = ref('');
    const loading = ref(false);
    const letterGroup = computed(() =>
      collectionName.value ? getLetterGroup(collectionName.value) : ''
    );
    const breadcrumbItems = computed(() => [
      {
        link: '/collections/A-J',
        text: 'Texts',
      },
      {
        link: `/collections/${letterGroup.value}`,
        text: letterGroup.value,
      },
    ]);
    const texts: Ref<CollectionText[]> = ref([]);
    const textsLoading = ref(false);
    const totalTexts = ref(0);
    const search = ref('');
    const page = ref(1);
    const rows = ref(10);

    const updateUrl = () => {
      let query: { page: string; rows: string; query?: string } = {
        page: String(page.value),
        rows: String(rows.value),
      };

      if (search.value !== '') {
        query.query = search.value;
      }

      if (router.currentRoute.name === 'collectionTexts') {
        router.replace({
          name: 'collectionTexts',
          params: {
            collectionUuid: props.collectionUuid,
          },
          query,
        });
      }
    };

    const getCollectionTexts = async () => {
      if (textsLoading.value) {
        return;
      }
      textsLoading.value = true;
      try {
        const collectionResp = await serverProxy.getCollectionTexts(
          props.collectionUuid,
          {
            page: page.value,
            rows: rows.value,
            query: search.value,
          }
        );
        totalTexts.value = collectionResp.totalTexts;
        texts.value = collectionResp.texts;
      } catch (err) {
        if (err.response && err.response.status === 403) {
          router.replace({ name: '403' });
        }
      } finally {
        textsLoading.value = false;
      }
    };

    watch(
      () => context.root.$route,
      () => {
        if (context.root.$route.name === 'collectionTexts') {
          getCollectionTexts();
        }
      }
    );

    onMounted(async () => {
      const {
        root: { $route },
      } = context;
      if ($route.query.page) {
        page.value = Number($route.query.page);
      } else {
        page.value = 1;
      }

      if ($route.query.rows) {
        rows.value = Number($route.query.rows);
      }

      if ($route.query.query) {
        search.value = String($route.query.query);
      } else {
        search.value = '';
      }

      loading.value = true;
      collectionName.value = (
        await serverProxy.getCollectionInfo(props.collectionUuid)
      ).name;
      loading.value = false;
    });

    watch(
      search,
      _.debounce(function() {
        page.value = 1;
        updateUrl();
      }, 500)
    );

    watch(page, () => {
      updateUrl();
    });

    watch(rows, () => {
      updateUrl();
    });

    return {
      collectionName,
      loading,
      texts,
      textsLoading,
      totalTexts,
      search,
      breadcrumbItems,
      page,
      rows,
    };
  },
});
</script>
