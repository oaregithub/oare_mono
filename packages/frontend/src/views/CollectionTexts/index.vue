<template>
  <v-progress-linear v-if="loading" indeterminate />
  <OareContentView v-else :title="collectionName">
    <template #header>
      <OareBreadcrumbs :items="breadcrumbItems" />
    </template>
    <v-spacer />
    <v-container>
      <v-row>
        <v-col cols="4" offset="8">
          <v-text-field
            :value="search"
            @input="(s) => setSearch(s || '')"
            label="Search"
            single-line
            hide-details
            clearable
            class="test-search"
          />
        </v-col>
      </v-row>
      <TextsTable
        :page="Number(page)"
        @update:page="(p) => setPage(String(p))"
        :rows="Number(rows)"
        @update:rows="(r) => setRows(String(r))"
        :totalTexts="totalTexts"
        :texts="texts"
        :loading="textsLoading"
      />
    </v-container>
  </OareContentView>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  watch,
  Ref,
  computed,
  onMounted,
  PropType,
} from '@vue/composition-api';
import Router from 'vue-router';
import { CollectionResponse, CollectionText } from '@/types/collections';
import { BreadcrumbItem } from '@/components/base/OareBreadcrumbs.vue';
import TextsTable from './TextsTable.vue';
import { getLetterGroup } from '../CollectionsView/utils';
import _ from 'underscore';
import defaultRouter from '@/router';
import defaultServerProxy from '@/serverProxy';
import useQueryParam from '@/hooks/useQueryParam';

export default defineComponent({
  name: 'CollectionTexts',
  components: {
    TextsTable,
  },
  props: {
    collectionUuid: {
      type: String,
      required: true,
    },
    router: {
      type: Object as PropType<Router>,
      default: () => defaultRouter,
    },
    serverProxy: {
      type: Object as PropType<typeof defaultServerProxy>,
      default: () => defaultServerProxy,
    },
    // useQueryParam: {
    //   type: Function as PropType<typeof useQueryParamDefault>,
    //   default: () => useQueryParamDefault,
    // },
  },

  setup(props, context) {
    const collectionName = ref('');
    const loading = ref(false);
    const letterGroup = computed(() =>
      collectionName.value ? getLetterGroup(collectionName.value) : ''
    );
    const breadcrumbItems = computed(() => [
      {
        link: '/collections/A-J',
        text: 'Texts',
      },
      {
        link: `/collections/${letterGroup.value}`,
        text: letterGroup.value,
      },
    ]);
    const texts: Ref<CollectionText[]> = ref([]);
    const textsLoading = ref(false);
    const totalTexts = ref(0);
    const [page, setPage] = useQueryParam('page', '1');
    const [rows, setRows] = useQueryParam('rows', '10');
    const [search, setSearch] = useQueryParam('query', '');

    const getCollectionTexts = async () => {
      if (textsLoading.value) {
        return;
      }
      textsLoading.value = true;
      try {
        const collectionResp = await props.serverProxy.getCollectionTexts(
          props.collectionUuid,
          {
            page: Number(page.value),
            rows: Number(rows.value),
            query: search.value,
          }
        );
        totalTexts.value = collectionResp.totalTexts;
        texts.value = collectionResp.texts;
      } catch (err) {
        if (err.response && err.response.status === 403) {
          props.router.replace({ name: '403' });
        }
      } finally {
        textsLoading.value = false;
      }
    };

    onMounted(async () => {
      loading.value = true;
      collectionName.value = (
        await props.serverProxy.getCollectionInfo(props.collectionUuid)
      ).name;
      loading.value = false;
    });

    watch([page, rows], () => {
      getCollectionTexts();
    });

    watch(
      search,
      _.debounce(function () {
        setPage('1');
        getCollectionTexts();
      }, 500),
      {
        lazy: true,
      }
    );

    return {
      collectionName,
      loading,
      texts,
      textsLoading,
      totalTexts,
      search,
      setSearch,
      breadcrumbItems,
      page,
      setPage,
      rows,
      setRows,
    };
  },
});
</script>
