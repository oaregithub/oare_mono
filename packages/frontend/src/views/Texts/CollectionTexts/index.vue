<template>
  <v-progress-linear v-if="loading" indeterminate />
  <OareContentView v-else :title="collectionName">
    <template #header v-if="!hideDetails">
      <OareBreadcrumbs :items="breadcrumbItems" />
    </template>
    <v-spacer />
    <v-container>
      <v-row>
        <v-col cols="4" offset="8">
          <v-text-field
            v-model="search"
            label="Search"
            single-line
            hide-details
            clearable
            class="test-search"
          />
        </v-col>
      </v-row>
      <v-btn v-if="canAddNewTexts" @click="addText" color="primary"
        >Add Text (BETA)</v-btn
      >
      <TextsTable
        :texts="visibleTexts"
        :loading="textsLoading"
        :collectionUuid="collectionUuid"
      />
    </v-container>
  </OareContentView>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  watch,
  Ref,
  computed,
  onMounted,
} from '@vue/composition-api';
import { CollectionText } from '@oare/types';
import TextsTable from './TextsTable.vue';
import { getLetterGroup } from '../CollectionsView/utils';
import _ from 'underscore';
import useQueryParam from '@/hooks/useQueryParam';
import sl from '@/serviceLocator';

export default defineComponent({
  name: 'CollectionTexts',
  components: {
    TextsTable,
  },
  props: {
    collectionUuid: {
      type: String,
      required: true,
    },
    hideDetails: {
      type: Boolean,
      default: false,
    },
  },

  setup({ collectionUuid }) {
    const server = sl.get('serverProxy');
    const actions = sl.get('globalActions');
    const router = sl.get('router');
    const store = sl.get('store');
    const hasBetaAccess = computed(() =>
      store.getters.user ? store.getters.user.betaAccess : false
    );

    const collectionName = ref('');
    const loading = ref(false);
    const letterGroup = computed(() =>
      collectionName.value ? getLetterGroup(collectionName.value) : ''
    );
    const breadcrumbItems = computed(() => [
      {
        link: '/collections/A-J',
        text: 'Texts',
      },
      {
        link: `/collections/${letterGroup.value}`,
        text: letterGroup.value,
      },
    ]);
    const texts: Ref<CollectionText[]> = ref([]);
    const visibleTexts: Ref<CollectionText[]> = ref([]);
    const textsLoading = ref(false);

    const search = useQueryParam('query', '', true);

    const addText = () => {
      router.push(`/add_collection_text/${collectionUuid}`);
    };

    const canAddNewTexts = computed(
      () => store.hasPermission('ADD_NEW_TEXTS') && hasBetaAccess.value
    );

    const getCollectionTexts = async () => {
      if (textsLoading.value) {
        return;
      }
      textsLoading.value = true;
      try {
        const collectionResp = await server.getCollectionTexts(collectionUuid);
        texts.value = collectionResp.texts;
        filterTexts();
      } catch (err) {
        actions.showErrorSnackbar(
          'Error loading collection texts. Please try again.',
          err as Error
        );
      } finally {
        textsLoading.value = false;
      }
    };

    onMounted(async () => {
      loading.value = true;
      try {
        collectionName.value = (
          await server.getCollectionInfo(collectionUuid)
        ).name;
        await getCollectionTexts();
      } catch (err) {
        actions.showErrorSnackbar(
          'Error loading collection texts. Please try again.',
          err as Error
        );
      } finally {
        loading.value = false;
      }
    });

    const filterTexts = () => {
      visibleTexts.value = texts.value.filter(text => {
        return (
          text.name.toLowerCase().includes(search.value.toLowerCase()) ||
          `${text.excavationPrefix || ''} ${text.excavationNumber || ''}`
            .toLowerCase()
            .includes(search.value.toLowerCase()) ||
          `${text.museumPrefix || ''} ${text.museumNumber || ''}`
            .toLowerCase()
            .includes(search.value.toLowerCase()) ||
          `${text.publicationNumber || ''} ${text.publicationNumber || ''}`
            .toLowerCase()
            .includes(search.value.toLowerCase())
        );
      });
    };

    watch(search, _.debounce(filterTexts, 100), {
      immediate: false,
    });

    return {
      collectionName,
      loading,
      texts,
      textsLoading,
      search,
      breadcrumbItems,
      addText,
      canAddNewTexts,
      visibleTexts,
    };
  },
});
</script>
