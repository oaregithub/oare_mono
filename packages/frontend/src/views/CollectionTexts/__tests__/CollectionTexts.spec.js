import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import { createLocalVue, mount } from '@vue/test-utils';
import CollectionTexts from '../index.vue';
import flushPromises from 'flush-promises';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('CollectionTexts test', () => {
  const mockTexts = [
    {
      id: 1,
      uuid: '1',
      type: 'text',
      hasEpigraphy: true,
      name: 'CCT 3 31',
    },
    {
      id: 2,
      uuid: '2',
      type: 'text',
      hasEpigraphy: true,
      name: 'CCT 1 12b',
    },
  ];

  const mockSetQueryParam = jest.fn();

  const mockProps = {
    collectionUuid: 'test-uuid',
    router: {
      replace: jest.fn(),
    },
    serverProxy: {
      getCollectionTexts: jest.fn().mockResolvedValue({
        collectionName: 'Test Collection',
        totalTexts: 2,
        texts: mockTexts,
      }),
      getCollectionInfo: jest.fn().mockResolvedValue({
        name: 'Test Collection',
      }),
    },
  };

  const createWrapper = async () => {
    const wrapper = mount(CollectionTexts, {
      propsData: mockProps,
      localVue,
      vuetify,
      stubs: ['router-link'],
    });
    await flushPromises();
    return wrapper;
  };

  it('matches snapshot', async () => {
    const wrapper = await createWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('updates URL when searching', async () => {
    const wrapper = await createWrapper();
    const searchInput = wrapper.find('.test-search input');
    searchInput.setValue('text name');
  });
});
