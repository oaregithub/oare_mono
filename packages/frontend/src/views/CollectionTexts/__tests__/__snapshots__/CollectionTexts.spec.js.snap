// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionTexts test matches snapshot 1`] = `
<div>
  <div class="d-flex">
    <div>
      <router-link-stub to="/collections/A-J" tag="a" ariacurrentvalue="page" event="click">Texts</router-link-stub> <span class="mx-3">&gt;</span>
    </div>
    <div>
      <router-link-stub to="/collections/L-Z" tag="a" ariacurrentvalue="page" event="click">L-Z</router-link-stub>
      <!---->
    </div>
  </div>
  <div class="v-card__title oare-header font-weight-bold d-flex justify-space-between align-start test-content-title">
    Test Collection
  </div>
  <div class="v-card__text black--text">
    <div class="spacer"></div>
    <div class="container">
      <div class="row">
        <div class="col col-4 offset-8">
          <div class="v-input test-search v-input--hide-details theme--light v-text-field v-text-field--single-line">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-34" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-34" type="text"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"></button></div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-table theme--light">
        <div class="v-data-table__wrapper">
          <table>
            <colgroup>
              <col class="">
            </colgroup>
            <thead class="v-data-table-header">
              <tr>
                <th role="columnheader" scope="col" aria-label="Text Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Text Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              </tr>
            </thead>
            <tbody>
              <tr class="">
                <td class="text-start">
                  <router-link-stub to="/epigraphies/1" tag="a" ariacurrentvalue="page" event="click">CCT 3 31</router-link-stub>
                </td>
              </tr>
              <tr class="">
                <td class="text-start">
                  <router-link-stub to="/epigraphies/2" tag="a" ariacurrentvalue="page" event="click">CCT 1 12b</router-link-stub>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div class="v-data-footer">
          <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-45" class="v-input__slot">
                  <div class="v-select__slot">
                    <div class="v-select__selections">
                      <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="$vuetify.dataTable.itemsPerPageText" id="input-45" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                    </div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden" value="10">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-data-footer__pagination">1-2 of 2</div>
          <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
          <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-card__actions"></div>
</div>
`;
